### ISSUE TITLE
define a hydra without a macro / lazily define a hydra

### ISSUE THREAD

**Issue by fommil** [abo-abo/hydra#149]
I'd like to be able to define a hydra, e.g.

```elisp
 (defhydra hydra-splitter (global-map "C-M-s")
  "splitter"
  ("<left>" hydra-move-splitter-left)
  ("<down>" hydra-move-splitter-down)
  ("<up>" hydra-move-splitter-up)
  ("<right>" hydra-move-splitter-right))
```

in such a way that I can choose to bind it to a function name and key later on instead of at the point of construction.

Basically, I lazily load all my packages https://github.com/fommil/dotfiles/blob/master/.emacs.d/init.el#L108 and by defining hydras via a macro, I can't see any way to lazily define hydras because the `defhydra` must be defined for my `init.el` to be valid, which means I have to aggressively load hydra. Alternatives to lazily defining hydras welcome.

**Comment by fommil** [comment]
I think this mostly does it... but it's pretty nasty. Alternative suggestions very welcome.

```elisp
(defun hydra-splitter/body ()
  "Defines a Hydra to resize the windows."
  ;; https://github.com/abo-abo/hydra/issues/149
  (interactive)
  (require 'hydra)
  (require 'hydra-examples)
  ;; overwrites the original keymapping and function
  (defhydra hydra-splitter (global-map "C-M-s")
    "splitter"
    ("<left>" hydra-move-splitter-left)
    ("<down>" hydra-move-splitter-down)
    ("<up>" hydra-move-splitter-up)
    ("<right>" hydra-move-splitter-right))
  (hydra-splitter/body))

(global-set-key (kbd "C-M-s") 'hydra-splitter/body)
```

**Comment by abo-abo** [comment]
> but it's pretty nasty

It sure is. There's no good way around it, and I don't think there should be: `defhydra` is a macro - it needs to be required. hydra.el isn't heavyweight at all - only 1000 lines.

If you really want to lazy load, here's what I suggest:

1. Move the definition of `hydra-splitter` into `hydras.el`.
2. You can't use the `global-map` approach: instead issue manual `global-set-key` in your `init.el`.
3. Add autoload statements to your `init.el`.

So in your init.el:

```elisp
(define-key global-map [134217747 left]
  'hydra-splitter/hydra-move-splitter-left)
(define-key global-map [134217747 down]
  'hydra-splitter/hydra-move-splitter-down)
(define-key global-map [134217747 up]
  'hydra-splitter/hydra-move-splitter-up)
(define-key global-map [134217747 right]
  'hydra-splitter/hydra-move-splitter-right)

(autoload 'hydra-splitter/hydra-move-splitter-left "hydras.el" "" t)
(autoload 'hydra-splitter/hydra-move-splitter-down "hydras.el" "" t)
(autoload 'hydra-splitter/hydra-move-splitter-up "hydras.el" "" t)
(autoload 'hydra-splitter/hydra-move-splitter-right "hydras.el" "" t)
```

In hydras.el:

    (require 'hydra)
    (require 'hydra-examples)
    ...

Basically you can require everything you need in hydras.el, since this file is autoloaded.  In the end, you have to decide if it's worth it - your original `defhydra` is 6 lines that expands to 250 lines.  You can add 12 extra lines to autoload the 250 lines.

**Comment by abo-abo** [comment]
Actually, you **can** avoid loading `hydra.el` with this approach: just remember load it when byte-compiling `hydras.el`.

**Comment by fommil** [comment]
thanks for the suggestions. I'll ponder it for a little longer. But I'm settling on this as a general pattern (`required` is my magical melpa lazy loader variant of `require`)

```elisp
(required 'hydra)
(defun hydra-splitter/body ()
  "Defines a Hydra to resize the windows."
  ;; overwrites the original function and calls it
  ;; https://github.com/abo-abo/hydra/issues/149
  (interactive)
  (require 'hydra-examples)
  (funcall
   (defhydra hydra-splitter nil "splitter"
     ("<left>" hydra-move-splitter-left)
     ("<down>" hydra-move-splitter-down)
     ("<up>" hydra-move-splitter-up)
     ("<right>" hydra-move-splitter-right))))

(global-set-key (kbd "C-M-s") 'hydra-splitter/body)
```

This allows me to:

1. lazily load `hydra`
2. define my global keybindings in the same place
3. not introduce too much boilerplate

at the expense of the ugliness of redefining the function as a side effect. I can live with that.
